flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/tableImpl.scala

```diff
@@ -313,135 +277,124 @@ class TableImpl(
   }
 
   override def unionAll(right: Table): Table = {
     // check that right table belongs to the same TableEnvironment
+    val rightImpl = right.asInstanceOf[TableImpl]
     if (right.asInstanceOf[TableImpl].tableEnv != this.tableEnv) {
       throw new ValidationException("Only tables from the same TableEnvironment can be unioned.")
     }
-    new TableImpl(
-      tableEnv, Union(logicalPlan, right.asInstanceOf[TableImpl].logicalPlan, all = true)
-        .validate(tableEnv))
+
+    new TableImpl(tableEnv,
+      operationTreeBuilder.union(logicalPlan, rightImpl.logicalPlan, all = true))
   }
 
   override def intersect(right: Table): Table = {
     // check that right table belongs to the same TableEnvironment
-    if (right.asInstanceOf[TableImpl].tableEnv != this.tableEnv) {
+    val rightImpl = right.asInstanceOf[TableImpl]
+    if (rightImpl.tableEnv != this.tableEnv) {
       throw new ValidationException(
         "Only tables from the same TableEnvironment can be intersected.")
     }
-    new TableImpl(
-      tableEnv, Intersect(logicalPlan, right.asInstanceOf[TableImpl].logicalPlan, all = false)
-        .validate(tableEnv))
+
+    new TableImpl(tableEnv,
+      operationTreeBuilder.intersect(logicalPlan, rightImpl.logicalPlan, all = false))
   }
 
   override def intersectAll(right: Table): Table = {
     // check that right table belongs to the same TableEnvironment
-    if (right.asInstanceOf[TableImpl].tableEnv != this.tableEnv) {
+    val rightImpl = right.asInstanceOf[TableImpl]
+    if (rightImpl.tableEnv != this.tableEnv) {
       throw new ValidationException(
         "Only tables from the same TableEnvironment can be intersected.")
     }
-    new TableImpl(
-      tableEnv, Intersect(logicalPlan, right.asInstanceOf[TableImpl].logicalPlan, all = true)
-        .validate(tableEnv))
+
+    new TableImpl(tableEnv,
+      operationTreeBuilder.intersect(logicalPlan, rightImpl.logicalPlan, all = true))
   }
 
   override def orderBy(fields: String): Table = {
     orderBy(ExpressionParser.parseExpressionList(fields).asScala: _*)
   }
 
   override def orderBy(fields: Expression*): Table = {
-    orderByInternal(fields.map(expressionBridge.bridge))
+    orderByInternal(fields)
   }
 
-  private def orderByInternal(fields: Seq[PlannerExpression]): Table = {
-    val order: Seq[Ordering] = fields.map {
-      case o: Ordering => o
-      case e => Asc(e)
-    }
-    new TableImpl(tableEnv, Sort(order, logicalPlan).validate(tableEnv))
+  private def orderByInternal(fields: Seq[Expression]): Table = {
+    new TableImpl(tableEnv,
+      operationTreeBuilder.sort(fields.map(_.accept(callResolver)).asJava, logicalPlan))
   }
 
   override def offset(offset: Int): Table = {
-    new TableImpl(tableEnv, Limit(offset, -1, logicalPlan).validate(tableEnv))
+    new TableImpl(tableEnv,
+      operationTreeBuilder.limit(Optional.of(offset), Optional.empty(), logicalPlan))
```

Comment:
Should we add `.limitWithOffset`, `.limitWithFetch`, `.limit()` instead of optionals?